pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -260,9 +260,14 @@ func (p *testPlanner) Plan() *TestPlan {
 		}
 	}
 
+	// Use first tested upgrade version in the context passed to startup
+	// functions; the setup upgrades that happened before should be
+	// invisible to them.
+	firstTestedUpgradeVersion := testUpgrades[0].from
```

Comment:
I am just trying to understand that there is no possibility of the length of testUpgrades to be 0 because the latest versions have webhooks?

Reply:
`testUpgrades` would be empty if the test's `minimumSupportedVersion` corresponds to a cluster version less than or equal to the the final build (i.e., the release where the test is running).

We validate[^1] that the minimum supported version is at least one release series older than the current build, so `testUpgrades` should never be empty.

[^1]: https://github.com/cockroachdb/cockroach/blob/8455f1b6714329fc2d7d2b57621b83d71c0e6578/pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go#L942-L957